# MongoDB
- https://www.mongodb.com/
- https://docs.mongodb.com/manual/tutorial/getting-started/
- https://www.w3schools.com/nodejs/nodejs_mongodb.asp
- NoSQL DB
- json-based document database (unlike traditional row/column model used by SQL)
- Rich JSON documents are the most natural productive way to work with data
- supports arrays and nested objects as values
- allows for flexible and dynamic schemas
- supports aggregations and modern use-cases such as geo-based search, graph search and text search
- queries themselves are JSON; just know your JSON!
- Learn basics from MongoDB University - course M001
    - create an account and learn for free; get certification at the end

## Schema validation
- collection schema validation for insert and update is allowed since MongoDB version 3.2
- see for details: https://docs.mongodb.com/manual/core/schema-validation/

## Atlas - MongoDB Cloud Database
- signup for free: https://www.mongodb.com/cloud/atlas/signup
- create a New Organization: `test-org`
- create a New Project: `test-project`
- create free Shared Clusters
    - pick AWS - North Aemerican region
    - give `test-cluster` name
    - takes 1-3 minutes to create the cluster
- create Database Access
    - select Password
    - under Password Authentication:
        - username: test-user
        - password: click Augogenerate Secure Password and Copy
    - under Database User Privileges: select Atlas admin
    - Add User
    
## Connect to Atlas cluster
- Under Clusters click CONNECT on test-cluster
    - Allow Access from Anywhere
        - Add default 0.0.0.0/0 IP Address
- Choose a Connection Method
- Select Connect your application
- copy connection string 
    - replace `<password>` with the `password` for the test-user
    - replace `myFirstDatabase` with the `database name` say `test-db`

## Connect to the server using NodeJS mongodb client
- install mongodb client driver for nodejs
- in order to properly work with Atlas MongoDB, you need to create `async functions` and call db methods with `await` keyword
- see await and async keyword descriptions here: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/async_function

```bash
cd <project_directory>
npm install mongodb
```

In [1]:
const MongoClient = require('mongodb').MongoClient;
// replace url with the connection string copied from Atlas cluster CONNECT step
// see Connect to Atlas cluster section above
// replace <password> with test-user's password
const db_user='paste your db username here'
const password='paste your password here'
const url = `mongodb+srv://${db_user}:${password}@cluster0.qvbyy.mongodb.net/myFirstDatabase?retryWrites=true&w=majority`
const db_options = {useNewUrlParser: true, useUnifiedTopology: true}

In [2]:
let db_name = 'test-db';
let coll_name = 'inventory';

In [3]:
// must use async function to connect; wait for certain operations 
// before closing the database
async function testConnection(url, db_options) { 
    // connect to cluster
    const client = await MongoClient.connect(url, db_options);
    client.close();
    console.log(`Successfully connected to database ${db_name}`)
}

In [4]:
testConnection(url, db_options, db_name);

Successfully connected to database test-db


## MongoDB CRUD Operations
- C: create
- R: read/find/select
- U: update
- D: delete

## Create Operations

### create collection
SQL 
```sql
CREATE TABLE table_name (
    column1 datatype,
    column2 datatype,
    column3 datatype,
   ....
);
```

Mongodb

```javascriopt
db.createCollection(collectionName);
```

In [14]:
async function createCollection(url, db_name, coll_name) {
    // await must be used before async function that returns Promise
    // inside async function
    let client = await MongoClient.connect(url, db_options);
    // specify DB name
    let db = client.db(db_name)
    let result = await db.createCollection(coll_name)
    client.close()
    console.log(`Collection ${coll_name} created!`)
    return result
}

In [15]:
createCollection(url, db_name, coll_name);

MongoError: Collection already exists. NS: test-db.inventory

### insert document
SQL
```SQL
INSERT INTO table_name (column1, column2, column3, ...)
VALUES (value1, value2, value3, ...);
```

Mongodb
### insert a single document
```javascript
collection.insertOne( {key: value, key1: value1, ...} );
```

In [11]:
async function insertOne(url, db_name, coll_name, js_obj) {
    const client = await MongoClient.connect(url, db_options);
    const db = client.db(db_name);
    const coll = db.collection(coll_name);
    const result = await coll.insertOne(js_obj);
    // if await is omitted, you'll get an error!
    client.close();
    console.log(`Message: ${result}`);
}

In [12]:
var item = { 
    item: 'canvas',
    qty: 100,
    tags: ['cotton'],
    size: {h: 28, w:35.5, uom: 'cm'}
};

In [13]:
insertOne(url, db_name, coll_name, item);

Message: {"result":{"n":1,"opTime":{"ts":"6950043940116496385","t":2},"electionId":"7fffffff0000000000000002","ok":1,"$clusterTime":{"clusterTime":"6950043940116496385","signature":{"hash":"pNumyHey7csrJEZSmsokiyFLsKU=","keyId":"6947013909408645122"}},"operationTime":"6950043940116496385"},"connection":{"_events":{},"_eventsCount":4,"id":1,"address":"44.240.7.138:27017","bson":{},"socketTimeout":0,"host":"cluster0-shard-00-01.qvbyy.mongodb.net","port":27017,"monitorCommands":false,"closed":false,"destroyed":false,"lastIsMasterMS":59},"ops":[{"item":"canvas","qty":100,"tags":["cotton"],"size":{"h":28,"w":35.5,"uom":"cm"},"_id":"60738425108a0b6815570ddc"}],"insertedCount":1,"insertedId":"60738425108a0b6815570ddc","n":1,"opTime":{"ts":"6950043940116496385","t":2},"electionId":"7fffffff0000000000000002","ok":1,"$clusterTime":{"clusterTime":"6950043940116496385","signature":{"hash":"pNumyHey7csrJEZSmsokiyFLsKU=","keyId":"6947013909408645122"}},"operationTime":"6950043940116496385"}


### insert many documents
```javascript
collection.insertMany([ json_doc1, json_doc2, ... ]);
```

In [14]:
async function insertMany(url, db_name, coll_name, obj_list) {
    const client = await MongoClient.connect(url, db_options);
    const db = client.db(db_name);
    const coll = db.collection(coll_name);
    const result = await coll.insertMany(obj_list);
    client.close();
    console.log(`${result.insertedCount} document(s) inserted!`)
}

In [15]:
var obj_list = 
    [
      {
        item: 'journal',
        qty: 25,
        tags: ['blank', 'red'],
        size: { h: 14, w: 21, uom: 'cm' }
      },
      {
        item: 'mat',
        qty: 85,
        tags: ['gray'],
        size: { h: 27.9, w: 35.5, uom: 'cm' }
      },
      {
        item: 'mousepad',
        qty: 25,
        tags: ['gel', 'blue'],
        size: { h: 19, w: 22.85, uom: 'cm' }
      }
    ]

In [16]:
insertMany(url, db_name, coll_name, obj_list)

3 document(s) inserted!


## read operations

### select all documents in a collection
SQL
```sql
SELECT * FROM <table_name>;
```

Mongodb
- must convert find all result to array
- use empty query object {}

```javascript
collection.find({}).toArray();
```

In [17]:
async function selectAll() {
    const client = await MongoClient.connect(url, db_options);
    const db = client.db(db_name);
    const coll = db.collection(coll_name);
    const items = await coll.find({}).toArray();
    client.close();
    return items;
}

In [18]:
console.log(selectAll());

Promise { <pending> }


### select/find one document

SQL

```sql
SELECT * FROM <TABLE> WHERE primary_key = value;
SELECT * FROM <TABLE> LIMIT 1;
```

- return the first document from results
```javascript
collection.findOne({});
```

In [38]:
async function findOne() {
    const client = await MongoClient.connect(url, db_options);
    const db = client.db(db_name);
    const coll = db.collection(coll_name);
    const item = await coll.findOne({});
    client.close();
    console.log(item);
}

In [39]:
findOne();

{
  _id: 60601216475892850f34a64e,
  item: 'canvas',
  qty: 100,
  tags: [ 'cotton' ],
  size: { h: 28, w: 35.5, uom: 'cm' }
}


### projection - limit fields/attributes to return from a query
- SQL
```sql
SELECT column1, column2, ... FROM tableName;
```

- MongoDB
- https://docs.mongodb.com/manual/tutorial/project-fields-from-query-results/
- {field: 0/1} : 0 omit, 1 include in result
- doesn't work here, but works fine in Compass!

```javascript
collection.find({}, {field:0, field1:1, ...});
```

In [40]:
async function selectFields() {
    const client = await MongoClient.connect(url, db_options);
    const db = client.db(db_name);
    const coll = db.collection(coll_name);
    // {fieldname: 1, fieldname: 0}; 1 include (bydefault), 0 exclude
    const items = await coll.find({}, {_id:0, item:0}).toArray();
    client.close();
    console.log(items);
}

In [41]:
selectFields();

[
  {
    _id: 60601216475892850f34a64e,
    item: 'canvas',
    qty: 100,
    tags: [ 'cotton' ],
    size: { h: 28, w: 35.5, uom: 'cm' }
  },
  {
    _id: 60601220475892850f34a64f,
    item: 'journal',
    qty: 25,
    tags: [ 'blank', 'red' ],
    size: { h: 14, w: 21, uom: 'cm' }
  },
  {
    _id: 60601220475892850f34a650,
    item: 'mat',
    qty: 85,
    tags: [ 'gray' ],
    size: { h: 27.9, w: 35.5, uom: 'cm' }
  },
  {
    _id: 60601220475892850f34a651,
    item: 'mousepad',
    qty: 25,
    tags: [ 'gel', 'blue' ],
    size: { h: 19, w: 22.85, uom: 'cm' }
  },
  {
    _id: 60601ad2475892850f34a652,
    item: 'journal',
    qty: 25,
    tags: [ 'blank', 'red' ],
    size: { h: 14, w: 21, uom: 'cm' }
  },
  {
    _id: 60601ad2475892850f34a653,
    item: 'mat',
    qty: 85,
    tags: [ 'gray' ],
    size: { h: 27.9, w: 35.5, uom: 'cm' }
  },
  {
    _id: 60601ad2475892850f34a654,
    item: 'mousepad',
    qty: 25,
    tags: [ 'gel', 'blue' ],
    size: { h: 19, w: 22.85, uom: 'c

### filter the result using query object
SQL
```sql
SELECT * FROM inventory WHERE item = 'canvas'
```

Mongodb
- explictly create a query object and pass it to find()
- must convert find result to array
```javascript
{fieldName: value}
```
- filter and return all documents having a given fieldName = value

In [46]:
async function executeQuery(query) {
    const client = await MongoClient.connect(url, db_options);
    const db = client.db(db_name);
    const coll =  db.collection(coll_name);
    const items = await coll.find(query).toArray();
    console.log(items)
    client.close();
}

In [47]:
var query = {item: 'mat'};

In [48]:
executeQuery(query);

[
  {
    _id: 60601220475892850f34a650,
    item: 'mat',
    qty: 85,
    tags: [ 'gray' ],
    size: { h: 27.9, w: 35.5, uom: 'cm' }
  },
  {
    _id: 60601ad2475892850f34a653,
    item: 'mat',
    qty: 85,
    tags: [ 'gray' ],
    size: { h: 27.9, w: 35.5, uom: 'cm' }
  }
]


### filter with regular expression
SQL
```sql
SELECT * FROM inventory WHERE item LIKE 'm%'
```
Mongodb
- find only the documents/objects where the item starts with the letter "m" - /^m/
 ```javascript
 db.find({ fieldName: re })
 ```

In [49]:
var query = {item: /^m/ };

In [50]:
executeQuery(query);

[
  {
    _id: 60601220475892850f34a650,
    item: 'mat',
    qty: 85,
    tags: [ 'gray' ],
    size: { h: 27.9, w: 35.5, uom: 'cm' }
  },
  {
    _id: 60601220475892850f34a651,
    item: 'mousepad',
    qty: 25,
    tags: [ 'gel', 'blue' ],
    size: { h: 19, w: 22.85, uom: 'cm' }
  },
  {
    _id: 60601ad2475892850f34a653,
    item: 'mat',
    qty: 85,
    tags: [ 'gray' ],
    size: { h: 27.9, w: 35.5, uom: 'cm' }
  },
  {
    _id: 60601ad2475892850f34a654,
    item: 'mousepad',
    qty: 25,
    tags: [ 'gel', 'blue' ],
    size: { h: 19, w: 22.85, uom: 'cm' }
  }
]


### specify AND, OR conditions and comparisons
SQL
```sql
select * from inventory where item like 'm%' and qty >= 85;
```
Mongodb

- Mongodb provides named operators for comparisons and logical operations
- $gt:  >

- $gte: >= 

- $lt: <

- $lte: <=

- $ne: !=

- $or: OR

```javascript
{ fieldName: {$gt: value} }
{ fieldName: {$lte: value} }
$or : [{fieldName1 : value1}, {fieldName2: value2}]
// and
{ fieldName1 : value1, fieldName2: {$lte: value2} }
```



In [52]:
var query = {item: /^m/,
            qty: {$ne: 85}}; // AND

In [53]:
executeQuery(query);

[
  {
    _id: 60601220475892850f34a651,
    item: 'mousepad',
    qty: 25,
    tags: [ 'gel', 'blue' ],
    size: { h: 19, w: 22.85, uom: 'cm' }
  },
  {
    _id: 60601ad2475892850f34a654,
    item: 'mousepad',
    qty: 25,
    tags: [ 'gel', 'blue' ],
    size: { h: 19, w: 22.85, uom: 'cm' }
  }
]


In [54]:
var query = {$or: [{item: /^m/}, {item: /^c/}], // OR
            qty: {$gte: 85}}; // AND

In [55]:
executeQuery(query);

[
  {
    _id: 60601216475892850f34a64e,
    item: 'canvas',
    qty: 100,
    tags: [ 'cotton' ],
    size: { h: 28, w: 35.5, uom: 'cm' }
  },
  {
    _id: 60601220475892850f34a650,
    item: 'mat',
    qty: 85,
    tags: [ 'gray' ],
    size: { h: 27.9, w: 35.5, uom: 'cm' }
  },
  {
    _id: 60601ad2475892850f34a653,
    item: 'mat',
    qty: 85,
    tags: [ 'gray' ],
    size: { h: 27.9, w: 35.5, uom: 'cm' }
  }
]


### sort the result - non-decreasing order
SQL
```sql
SELECT column1, column2, ...
FROM table_name
ORDER BY column1, column2, ... ASC;
```

Mongodb
- sort in ascending (non-decreasing) order based on fieldName
```javascript
collection.find({}).sort({ fieldName: 1 })
```

In [57]:
async function getSortedResults(query, sort_option) {
    const client = await MongoClient.connect(url, db_options);
    const db = client.db(db_name);
    const coll =  db.collection(coll_name);
    const items = await coll.find(query).sort(sort_option).toArray();
    console.log(items)
    client.close();
}

In [58]:
var query = {$or: [{item: /^m/}, {item: /^c/}], // OR
             qty: {$gte: 85}}; // AND
var sort_options = {qty: 1}; // Ascending order

In [59]:
getSortedResults(query, sort_options)

[
  {
    _id: 60601220475892850f34a650,
    item: 'mat',
    qty: 85,
    tags: [ 'gray' ],
    size: { h: 27.9, w: 35.5, uom: 'cm' }
  },
  {
    _id: 60601ad2475892850f34a653,
    item: 'mat',
    qty: 85,
    tags: [ 'gray' ],
    size: { h: 27.9, w: 35.5, uom: 'cm' }
  },
  {
    _id: 60601216475892850f34a64e,
    item: 'canvas',
    qty: 100,
    tags: [ 'cotton' ],
    size: { h: 28, w: 35.5, uom: 'cm' }
  }
]


### sort the result - non-increasing order
SQL
```sql
SELECT column1, column2, ...
FROM table_name
ORDER BY column1, column2, ... ASC|DESC;
```

Mongodb

- sort the result in descending (non-increasing) order
```javascript
collection.find({}).sort({ fieldName: -1 })
```

In [60]:
// sort in descending order
getSortedResults(query, {qty: -1});

[
  {
    _id: 60601216475892850f34a64e,
    item: 'canvas',
    qty: 100,
    tags: [ 'cotton' ],
    size: { h: 28, w: 35.5, uom: 'cm' }
  },
  {
    _id: 60601220475892850f34a650,
    item: 'mat',
    qty: 85,
    tags: [ 'gray' ],
    size: { h: 27.9, w: 35.5, uom: 'cm' }
  },
  {
    _id: 60601ad2475892850f34a653,
    item: 'mat',
    qty: 85,
    tags: [ 'gray' ],
    size: { h: 27.9, w: 35.5, uom: 'cm' }
  }
]


## update document
SQL
```sql
UPDATE table_name
SET column1 = value1, column2 = value2, ...
WHERE condition;
```

### updateOne(query, newValue, callback)

- use $set operator to set {key : newvalue}

```javascript
collection.updateOne( {query}, {$set: {key : newvalue}} );
```
- only 1 document (the first matching) is updated even if the query matches many documents

In [67]:
async function updateOne(query, newVal) {
    const client = await MongoClient.connect(url, db_options);
    const db = client.db(db_name);
    const coll = db.collection(coll_name);
    const res = await coll.updateOne(query, newValue);
    console.log(`${res.result.nModified} document updated!`);
    client.close();
}

In [68]:
var query = { item: /^m/,
            qty: {$gte: 85}};
var newValue = {$set: {qty: 75}}

In [69]:
updateOne(query, newValue);

1 document updated!


### updateMany(query, newValue, callback) 
- update all documents that meet the criteria of the query
- use $set operator to set the { key : newvalue}

```javascript
collection.updateMany({query}, {$set: {key: newValue})
```

In [71]:
async function updateMany(query, newVal) {
    const client = await MongoClient.connect(url, db_options);
    const db = client.db(db_name);
    const coll = db.collection(coll_name);
    const res = await coll.updateMany(query, newValue);
    console.log(`${res.result.nModified} document(s) updated!`);
    client.close();
}

In [73]:
var query = { item: /^m/,
             qty: {$gte: 50}};
var newValue = {$set: {qty: 70}};


In [74]:
updateMany(query, newValue);

2 document(s) updated!


## delete operations

### delete document
SQL
```sql
DELETE FROM table_name WHERE condition;
```

Mongodb

### deleteOne(query, callback)
- delete a record/document
- if query matches multiple documents, first occurance is deleted
```javascript
collection.deleteOne({query});
```

In [76]:
async function deleteOne (query) {
    const client = await MongoClient.connect(url, db_options);
    const db = client.db(db_name);
    const coll = db.collection(coll_name);
    const res = await coll.deleteOne(query);
    console.log(`${res.result.n} document deleted!`);
    client.close();
}

In [77]:
var query = { item: /^m/,
                 qty: {$gte: 70}};

In [78]:
deleteOne(query);

1 document deleted!


### deleteMany(query, callback)
- delete all records/documents matching query
- careful with this; can't undo!

```javascript
db.deleteMany({query}, ()=>{})
```

In [79]:
async function deleteMany (query) {
    const client = await MongoClient.connect(url, db_options);
    const db = client.db(db_name);
    const coll = db.collection(coll_name);
    const res = await coll.deleteMany(query);
    console.log(`${res.result.n} document deleted!`);
    client.close();
}

In [80]:
var query = { item: /^m/,
             qty: 25};

In [81]:
deleteMany(query);

2 document deleted!


### drop collection
- careful with drop; can't undo!

SQL
```
DROP TABLE tableName
```

Mongodb
```javascript
db.dropCollection(collectionName);
```

In [83]:
async function dropCollection(coll_name) {
    const client = await MongoClient.connect(url, db_options);
    const db = client.db(db_name);
    const res = await db.dropCollection(coll_name);
    console.log(`${res} Collection dropped!`);
    client.close();
}

In [84]:
dropCollection(coll_name);

true Collection dropped!


## MongoDB GUI-based clients and management tools
- MongoDB Compass Community Edition: https://www.mongodb.com/download-center/compass

## Install MongoDB
- https://docs.mongodb.com/manual/administration/install-community/
- E.g., on Mac, you can use brew to install Mongodb
    - see: https://docs.mongodb.com/manual/tutorial/install-mongodb-on-os-x/

## Run MongoDB local server
- add Mongodb's /bin folder to path or simply CD into the folder to run mongod app
- Using terminal: 
    - $ mongod --dbpath=[dbfolder]
    
- CD into MongoDB folder and run the following command:
    - $ mongod --dbpath=data
- MongoDB by defualt runs on localhost:27017

## Connect to the local/remote Mongo server
- you can use CLI mongo client or Compass GUI-based client
<img src="./resources/MongoDBCompass.png">